api/handlers/listhandler.py

```diff
@@ -480,13 +480,67 @@ def get(self, cont_name, list_name, **kwargs):
 
         # Authenticated or ticketed download request
         else:
-            self.response.app_iter = open(filepath, 'rb')
-            self.response.headers['Content-Length'] = str(fileinfo['size']) # must be set after setting app_iter
-            if self.is_true('view'):
-                self.response.headers['Content-Type'] = str(fileinfo.get('mimetype', 'application/octet-stream'))
+            range_header = self.request.headers.get('Range', '')
+            try:
+                if not self.is_true('view'):
+                    raise util.RangeHeaderParseError('Feature flag not set')
+
+                ranges = util.parse_range_header(range_header)
+                for first, last in ranges:
+                    if first > fileinfo['size'] - 1:
+                        self.abort(416, 'Invalid range')
+
+                    if last > fileinfo['size'] - 1:
+                        raise util.RangeHeaderParseError('Invalid range')
+
+            except util.RangeHeaderParseError:
+                self.response.app_iter = open(filepath, 'rb')
+                self.response.headers['Content-Length'] = str(fileinfo['size'])  # must be set after setting app_iter
+
+                if self.is_true('view'):
+                    self.response.headers['Content-Type'] = str(fileinfo.get('mimetype', 'application/octet-stream'))
+                else:
+                    self.response.headers['Content-Type'] = 'application/octet-stream'
+                    self.response.headers['Content-Disposition'] = 'attachment; filename="' + filename + '"'
             else:
-                self.response.headers['Content-Type'] = 'application/octet-stream'
-                self.response.headers['Content-Disposition'] = 'attachment; filename="' + filename + '"'
+                self.response.status = 206
```

Comment:
Are we sure `try - except - else` is a pattern we want to use here? It feels weird to grab the whole stack & throw `Feature flag not set` to fill the role of an if statement.

Reply:
We want to "fail back" to serving the whole file for any parsing issue. I think in this case we consider the feature flag to be part of our modified RFC7233 protocol. If we didn't make that assumption, we would still need an if statement outside a try/except block where the except executes the same code. That seems even more confusing.